File: go/internal/core/list.go
package core

import (
	"bytes"
	"encoding/json"
	"io/fs"
	"path"
	"sort"
	"strings"
)

// SkillInfo describes a discovered skill for listing.
type SkillInfo struct {
	ID      string
	Root    Root
	Summary string
	Version string
}

// SkillListJSONItem preserves the exact key order of the reference output.
type SkillListJSONItem struct {
	ID      string `json:"id"`
	Digest  string `json:"digest"`
	Files   int    `json:"files,omitempty"`
	Summary string `json:"summary,omitempty"`
	Version string `json:"version,omitempty"`
}

// SkillListJSON is the payload of `--skill list --json`.
type SkillListJSON struct {
	SkillflagVersion string              `json:"skillflag_version"`
	Skills           []SkillListJSONItem `json:"skills"`
}

var summaryReplacer = strings.NewReplacer("\t", " ", "\n", " ")

func readSkillInfo(root Root, id string) SkillInfo {
	info := SkillInfo{ID: id, Root: root}
	content, err := fs.ReadFile(root.FS, path.Join(id, "SKILL.md"))
	if err != nil {
		return info
	}
	fields := ParseFrontmatter(string(content))
	if description := fields["description"]; description != "" {
		info.Summary = strings.TrimSpace(summaryReplacer.Replace(description))
	}
	info.Version = fields["version"]
	return info
}

// ListSkills merges skills from every root (first root wins on duplicate ids)
// and returns them sorted byte-wise by id.
func ListSkills(roots []Root) []SkillInfo {
	type picked struct {
		id   string
		root Root
	}
	seen := map[string]bool{}
	var picks []picked
	for _, root := range roots {
		for _, ref := range ListSkillDirs(root) {
			if !seen[ref.ID] {
				seen[ref.ID] = true
				picks = append(picks, picked{id: ref.ID, root: root})
			}
		}
	}

	infos := make([]SkillInfo, 0, len(picks))
	for _, p := range picks {
		infos = append(infos, readSkillInfo(p.root, p.id))
	}
	sort.Slice(infos, func(i, j int) bool { return infos[i].ID < infos[j].ID })
	return infos
}

// ListSkillsJSON builds the JSON listing including per-skill digests.
func ListSkillsJSON(roots []Root) (SkillListJSON, error) {
	skills := ListSkills(roots)
	items := make([]SkillListJSONItem, 0, len(skills))

	for _, skill := range skills {
		skillFS, subErr := fs.Sub(skill.Root.FS, skill.ID)
		if subErr != nil {
			return SkillListJSON{}, subErr
		}
		entries, fileCount, err := CollectSkillEntries(skillFS, skill.ID, skill.Root.NormalizeModes)
		if err != nil {
			return SkillListJSON{}, err
		}
		digest, err := DigestSkill(skillFS, entries)
		if err != nil {
			return SkillListJSON{}, err
		}
		items = append(items, SkillListJSONItem{
			ID:      skill.ID,
			Digest:  digest,
			Files:   fileCount,
			Summary: skill.Summary,
			Version: skill.Version,
		})
	}

	return SkillListJSON{SkillflagVersion: "0.1", Skills: items}, nil
}

// MarshalListJSON renders the payload as compact JSON without HTML escaping
// and without a trailing newline, matching JSON.stringify byte-for-byte.
func MarshalListJSON(payload SkillListJSON) ([]byte, error) {
	var buf bytes.Buffer
	encoder := json.NewEncoder(&buf)
	encoder.SetEscapeHTML(false)
	if err := encoder.Encode(payload); err != nil {
		return nil, err
	}
	return bytes.TrimSuffix(buf.Bytes(), []byte("\n")), nil
}

File: rust/crates/skillflag/src/core/list.rs
use std::fs;

use crate::bundled;
use crate::core::digest::digest_sha256;
use crate::core::errors::Result;
use crate::core::paths::{list_skill_dirs, SkillDirRef, SkillsRoot};
use crate::core::tar::{collect_skill_entries, write_tar};
use crate::frontmatter::parse_frontmatter;

#[derive(Debug, Clone)]
pub struct SkillInfo {
    pub id: String,
    pub dir: SkillDirRef,
    pub summary: Option<String>,
    pub version: Option<String>,
}

fn skill_md_text(dir: &SkillDirRef) -> Option<String> {
    match dir {
        SkillDirRef::Disk(path) => fs::read_to_string(path.join("SKILL.md")).ok(),
        SkillDirRef::Bundled(id) => bundled::skill_md(id).map(str::to_string),
    }
}

fn read_skill_info(id: String, dir: SkillDirRef) -> SkillInfo {
    let Some(content) = skill_md_text(&dir) else {
        return SkillInfo {
            id,
            dir,
            summary: None,
            version: None,
        };
    };
    let fields = parse_frontmatter(&content);
    let summary = fields
        .get("description")
        .map(|d| d.replace(['\t', '\n'], " ").trim().to_string())
        .filter(|s| !s.is_empty());
    let version = fields.get("version").cloned().filter(|v| !v.is_empty());
    SkillInfo {
        id,
        dir,
        summary,
        version,
    }
}

/// List skills across roots, sorted byte-wise by id. When the same id exists
/// in several roots, the FIRST root wins.
pub fn list_skills(roots: &[SkillsRoot]) -> Vec<SkillInfo> {
    let mut seen: Vec<(String, SkillDirRef)> = Vec::new();
    for root in roots {
        for (id, dir) in list_skill_dirs(root) {
            if !seen.iter().any(|(seen_id, _)| *seen_id == id) {
                seen.push((id, dir));
            }
        }
    }

    let mut infos: Vec<SkillInfo> = seen
        .into_iter()
        .map(|(id, dir)| read_skill_info(id, dir))
        .collect();
    infos.sort_by(|a, b| a.id.cmp(&b.id));
    infos
}

/// Escape a string the way `JSON.stringify` does.
fn json_string(value: &str) -> String {
    let mut out = String::with_capacity(value.len() + 2);
    out.push('"');
    for ch in value.chars() {
        match ch {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\u{8}' => out.push_str("\\b"),
            '\t' => out.push_str("\\t"),
            '\n' => out.push_str("\\n"),
            '\u{c}' => out.push_str("\\f"),
            '\r' => out.push_str("\\r"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
    out
}

/// Build the compact `list --json` payload (no trailing newline; fixed key
/// order: id, digest, files, summary, version; optional fields omitted).
pub fn list_skills_json(roots: &[SkillsRoot]) -> Result<String> {
    let skills = list_skills(roots);
    let mut items: Vec<String> = Vec::with_capacity(skills.len());
    for skill in &skills {
        let collected = collect_skill_entries(&skill.dir, &skill.id)?;
        let tar = write_tar(&collected.entries)?;
        let digest = digest_sha256(&tar);

        let mut item = format!(
            "{{\"id\":{},\"digest\":{}",
            json_string(&skill.id),
            json_string(&digest)
        );
        if collected.file_count > 0 {
            item.push_str(&format!(",\"files\":{}", collected.file_count));
        }
        if let Some(summary) = &skill.summary {
            item.push_str(&format!(",\"summary\":{}", json_string(summary)));
        }
        if let Some(version) = &skill.version {
            item.push_str(&format!(",\"version\":{}", json_string(version)));
        }
        item.push('}');
        items.push(item);
    }
    Ok(format!(
        "{{\"skillflag_version\":\"0.1\",\"skills\":[{}]}}",
        items.join(",")
    ))
}

#[cfg(test)]
mod tests {
    use super::json_string;

    #[test]
    fn json_string_escapes_like_json_stringify() {
        assert_eq!(json_string("plain"), "\"plain\"");
        assert_eq!(json_string("a\"b\\c"), "\"a\\\"b\\\\c\"");
        assert_eq!(
            json_string("tab\tnl\ncr\rbs\u{8}ff\u{c}"),
            "\"tab\\tnl\\ncr\\rbs\\bff\\f\""
        );
        assert_eq!(json_string("\u{1}\u{1f}"), "\"\\u0001\\u001f\"");
        assert_eq!(json_string("ünïcode"), "\"ünïcode\"");
    }
}
